# Case Study: Analyzing U.S. Storm Events for Disaster Preparedness 

In this case study, we will analyze storm event data to enhance disaster preparedness efforts in the U.S. We will follow a structured data analysis workflow comprising the following phases:

1. **Ask**: Define the problem and confirm expectations.
2. **Prepare**: Collect and store data for analysis.
3. **Process**: Clean and transform data to ensure integrity.
4. **Analyze**: Use data analysis tools to draw conclusions.
5. **Share**: Interpret and communicate results to make data-driven decisions.
6. **Act**: Put insights to work to address the original problem.

We will utilize the open-source Storm Event Database provided by the National Oceanic and Atmospheric Administration (NOAA), available at [NOAA Storm Events Database](https://www.ncdc.noaa.gov/stormevents/ftp.jsp). This dataset contains comprehensive information about various storm events across the U.S., including details such as event type, location, and impact.

## Ask

### Problem Statement
The U.S. experiences various storm events that can impact public safety, property, and infrastructure. Analyzing the patterns and impacts of these storms will provide insights that can inform disaster preparedness and response strategies.

### Key Questions
1. **What types of storm events are most prevalent in the U.S.?**  
   Identify the most common storm events and their geographical distribution.

2. **Where do storm events most frequently occur?**  
   Map the geographical distribution of storm events to identify high-risk areas.

3. **Are certain regions more prone to specific types of storm events?**  
   Analyze the geographic hot spots for storm types to develop tailored regional preparedness plans.

4. **What is the temporal distribution of storm events?**  
   Analyze trends over time to understand seasonal patterns and changes in frequency.

5. **What is the impact of different storm events on injuries, fatalities, and property damage?**  
   Assess the severity of various storm types and their associated risks to prioritize high-risk events for preparedness planning.

6. **How can this analysis inform disaster preparedness initiatives?**  
   Determine actionable insights that can help stakeholders enhance response plans and allocate resources effectively.

## Prepare

### Data Sources
We will use the Storm Events Database from the NOAA for the years 2014 to 2024. The dataset can be accessed at [NOAA Storm Events Database](https://www.ncdc.noaa.gov/stormevents/ftp.jsp).

### Data Collection
The relevant files will be downloaded in CSV.GZ format, which is a compressed version of CSV files.

### Data Description
The dataset contains 51 columns with various types of information related to storm events. Below is the data dictionary presented in a table format:


| Column Name         | Example                                                    | Description                                                                                                             |
|:--------------------|:-----------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------|
| begin_yearmonth     | 201212                                                     | The year and month that the event began (YYYYMM format).                                                                |
| begin_day           | 31                                                         | The day of the month that the event began (DD format).                                                                  |
| begin_time          | 2359                                                       | The time of day that the event began (hhmm format).                                                                     |
| end_yearmonth       | 201301                                                     | The year and month that the event ended (YYYYMM format).                                                                |
| end_day             | 01                                                         | The day of the month that the event ended (DD format).                                                                  |
| end_time            | 0001                                                       | The time of day that the event ended (hhmm format).                                                                     |
| episode_id          | 61280                                                      | ID assigned by NWS to denote the storm episode; may contain multiple events.                                            |
| event_id            | 383097                                                     | ID assigned by NWS for each individual storm event (Primary database key field).                                        |
| state               | GEORGIA                                                    | The state name where the event occurred (ALL CAPS).                                                                     |
| state_fips          | 45                                                         | Unique number assigned to the county by NIST (State FIPS).                                                              |
| year                | 2000                                                       | The four-digit year for the event in this record.                                                                       |
| month_name          | January                                                    | The name of the month for the event (not abbreviated).                                                                  |
| event_type          | Hail                                                       | The type of storm event (spelled out; not abbreviated).                                                                 |
| cz_type             | C                                                          | Indicates whether the event happened in a County/Parish, NWS Public Forecast Zone, or Marine.                           |
| cz_fips             | 245                                                        | The county FIPS number assigned by NIST or NWS Forecast Zone Number.                                                    |
| cz_name             | AIKEN                                                      | Name assigned to the county FIPS number or NWS Forecast Zone.                                                           |
| wfo                 | CAE                                                        | The NWS Forecast Office’s area of responsibility in which the event occurred.                                           |
| begin_date_time     | 04/1/2012 20:48:00                                         | Start date and time of the event (MM/DD/YYYY hh:mm:ss).                                                                 |
| cz_timezone         | EST-5                                                      | Time Zone for the County/Parish, Zone or Marine Name.                                                                   |
| end_date_time       | 04/1/2012 21:03:00                                         | End date and time of the event (MM/DD/YYYY hh:mm:ss).                                                                   |
| injuries_direct     | 1                                                          | Number of injuries directly caused by the weather event.                                                                |
| injuries_indirect   | 0                                                          | Number of injuries indirectly caused by the weather event.                                                              |
| deaths_direct       | 0                                                          | Number of deaths directly caused by the weather event.                                                                  |
| deaths_indirect     | 0                                                          | Number of deaths indirectly caused by the weather event.                                                                |
| damage_property     | 10.00K                                                     | Estimated property damage incurred by the weather event.                                                                |
| damage_crops        | 0.00K                                                      | Estimated damage to crops incurred by the weather event.                                                                |
| source              | Public                                                     | Source reporting the weather event.                                                                                     |
| magnitude           | 0.75                                                       | Measured extent of the magnitude type (only for wind speeds and hail size).                                             |
| magnitude_type      | EG                                                         | Type of magnitude measurement (e.g., wind estimated gust).                                                              |
| flood_cause         | Ice Jam                                                    | Reported cause of the flood.                                                                                            |
| category            |                                                     | Unknown (During the time of downloading this particular file, NCDC has never seen anything provided within this field.) |
| tor_f_scale         | EF0                                                        | Enhanced Fujita Scale describing tornado strength.                                                                      |
| tor_length          | 0.66                                                       | Length of the tornado while on the ground (in miles).                                                                   |
| tor_width           | 25                                                         | Width of the tornado while on the ground (in whole yards).                                                              |
| tor_other_wfo       | DDC                                                        | Continuation of a tornado segment as it crossed from one NWS Forecast Office to another.                                |
| tor_other_cz_state  | KS                                                         | Two-character representation for the state name of the continuing tornado segment.                                      |
| tor_other_cz_fips   | 41                                                         | FIPS number of the county for the continuing tornado segment.                                                           |
| tor_other_cz_name   | DICKINSON                                                  | Name of the county for the continuing tornado segment.                                                                  |
| begin_range         | 0.59                                                       | Distance to the nearest tenth of a mile to the location referenced.                                                     |
| begin_azimuth       | ENE                                                        | 16-point compass direction from the location referenced.                                                                |
| begin_location      | PINELAND                                                   | Name of city, town, or village from which the range is calculated.                                                      |
| end_range           | 0.66                                                       | See begin_range.                                                                                                        |
| end_azimuth         | WNW                                                        | See begin_azimuth.                                                                                                      |
| end_location        | RUSK                                                       | See begin_location.                                                                                                     |
| begin_lat           | 29.7898                                                    | Latitude in decimal degrees of the begin point of the event.                                                            |
| begin_lon           | -98.6406                                                   | Longitude in decimal degrees of the begin point of the event.                                                           |
| end_lat             | 29.7158                                                    | Latitude in decimal degrees of the end point of the event.                                                              |
| end_lon             | -98.7744                                                   | Longitude in decimal degrees of the end point of the event.                                                             |
| episode_narrative   | A strong upper level system over the southern Rockies...   | Narrative depicting the general nature of the episode.                                                                  |
| event_narrative     | Heavy rain caused flash flooding across parts of Wilber... | Narrative providing descriptive details of the individual event.                                                        |

### Data Storage
The data will be stored in a local environment for analysis.

### Data Loading and Initial Exploration

In [128]:
# To begin, let's import all necessary libraries
import pandas as pd
import numpy as np
import janitor
import matplotlib.pyplot as plt
import seaborn as sns
import contextily as ctx
import glob
import missingno as msno

from shapely.geometry import Point
import geopandas as gpd
import folium

from scipy.stats import chi2_contingency

# Set display options to show all columns
pd.set_option('display.max_columns', None)

In [39]:
# In the data folder, we have lots of files. But for this analysis, we only need Storm Details files which have the prefix "StormEvents_details-ftp_v1.0_d".
# So, let's define the file path pattern
file_pattern = "./data/StormEvents_details-ftp_v1.0_d*.csv.gz"

# Get a list of all files matching the pattern
all_files = glob.glob(file_pattern)

# Read and concatenate all files into a single DataFrame
df_list = [pd.read_csv(file, compression='gzip') for file in all_files]
df_details = pd.concat(df_list, ignore_index=True)

# Based on the data dictionary, we can drop the following columns: `CATEGORY`
df_details.drop(columns=['CATEGORY'], inplace=True)

# Rename column names to snake case for consistency 
df_details = df_details.clean_names()

# Display the first few rows of the DataFrame
df_details.head()

begin_yearmonth  begin_day  begin_time  end_yearmonth  end_day  end_time  \
0           202112         11         349         202112       11       350   
1           202112         11         249         202112       11       254   
2           202112         11         325         202112       11       327   
3           202112         11         232         202112       11       239   
4           202112          6         724         202112        6       724   

   episode_id  event_id      state  state_fips  ...  end_range end_azimuth  \
0      165322    999750  TENNESSEE          47  ...        3.0          NW   
1      165322    999613  TENNESSEE          47  ...        2.0         ESE   
2      165322    999636  TENNESSEE          47  ...        2.0          SW   
3      165322    999604  TENNESSEE          47  ...        4.0         NNW   
4      165321    999306  TENNESSEE          47  ...        1.0           W   

  end_location begin_lat  begin_lon  end_lat  end_lon  \
0   HUNTERS PT   36.3178   -86.3235  36.3296 -86.2965   
1  BAKERSWORKS   36.0255   -87.3054  36.0736 -87.2330   
2        AMQUI   36.2372   -86.7286  36.2572 -86.7035   
3     PINEWOOD   35.9205   -87.6423  35.9725 -87.5068   
4    JAMESTOWN   36.4322   -84.9405  36.4322 -84.9405   

                                   episode_narrative  \
0  One of the worst tornado outbreaks ever record...   
1  One of the worst tornado outbreaks ever record...   
2  One of the worst tornado outbreaks ever record...   
3  One of the worst tornado outbreaks ever record...   
4  After some isolated thunderstorms moved across...   

                                     event_narrative data_source  
0  This small EF-0 tornado was determined through...         CSV  
1  This tornado developed just southeast of the D...         CSV  
2  Severe straight-line winds caused significant ...         CSV  
3  This tornado touched down in far northwest Hic...         CSV  
4  A Facebook report indicated trees and power li...         CSV  

[5 rows x 50 columns]

In [73]:
# Display the DataFrame statistics
df_details.describe()

begin_yearmonth      begin_day     begin_time  end_yearmonth  \
count    669746.000000  669746.000000  669746.000000  669746.000000   
mean     201909.724999      14.884967    1262.750553  201909.724999   
std         308.331723       9.061630     677.789666     308.331723   
min      201401.000000       1.000000       0.000000  201401.000000   
25%      201611.000000       7.000000     745.000000  201611.000000   
50%      201907.000000      15.000000    1434.000000  201907.000000   
75%      202204.000000      23.000000    1800.000000  202204.000000   
max      202406.000000      31.000000    2359.000000  202406.000000   

             end_day       end_time     episode_id      event_id  \
count  669746.000000  669746.000000  669746.000000  6.697460e+05   
mean       16.502822    1454.988749  138990.624232  8.409394e+05   
std         9.079046     615.305445   32771.003907  2.038179e+05   
min         1.000000       0.000000   80464.000000  4.819030e+05   
25%         9.000000    1051.000000  111238.000000  6.643362e+05   
50%        16.000000    1600.000000  139511.000000  8.409355e+05   
75%        24.000000    1900.000000  167882.000000  1.017476e+06   
max        31.000000    2359.000000  194241.000000  1.201899e+06   

          state_fips           year  ...      magnitude   tor_length  \
count  669746.000000  669746.000000  ...  348945.000000  15192.00000   
mean       33.071031    2019.038506  ...      37.897867      3.16236   
std        19.208950       3.084325  ...      23.715012      4.14433   
min         1.000000    2014.000000  ...       0.130000      0.01000   
25%        19.000000    2016.000000  ...       1.750000      0.50000   
50%        31.000000    2019.000000  ...      50.000000      1.66000   
75%        46.000000    2022.000000  ...      52.000000      4.20000   
max        99.000000    2024.000000  ...     173.000000     41.88000   

          tor_width  tor_other_cz_fips    begin_range      end_range  \
count  15192.000000        1988.000000  410429.000000  410429.000000   
mean     196.235153         105.344064       2.440858       2.463415   
std      296.644097          86.964479       4.563890       4.577299   
min        1.000000           1.000000       0.000000       0.000000   
25%       50.000000          45.000000       1.000000       1.000000   
50%      100.000000          91.000000       1.000000       1.000000   
75%      200.000000         143.000000       3.000000       3.000000   
max     3960.000000         820.000000     185.000000     185.000000   

           begin_lat      begin_lon        end_lat        end_lon  
count  410429.000000  410429.000000  410429.000000  410429.000000  
mean       37.648997     -90.233009      37.647509     -90.226771  
std         5.177361      11.722470       5.178816      11.720015  
min       -14.400000    -171.032700     -14.437500    -170.905900  
25%        34.310000     -97.260000      34.310000     -97.260000  
50%        38.140000     -89.530000      38.140000     -89.517600  
75%        41.220000     -81.460000      41.218300     -81.450000  
max        70.375400     151.848400      70.264600     151.858900  

[8 rows x 25 columns]

### Check for Data Quality Issues

In [40]:
# Now let check for duplicates in the data
duplicates = df_details.duplicated().sum()
print(f"Number of duplicate rows: {duplicates}")

Number of duplicate rows: 0


That's great news. We have no duplicated rows in our data. Let's proceed to check for missing values in the data.

In [41]:
# Display the DataFrame information
df_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 669746 entries, 0 to 669745
Data columns (total 50 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   begin_yearmonth     669746 non-null  int64  
 1   begin_day           669746 non-null  int64  
 2   begin_time          669746 non-null  int64  
 3   end_yearmonth       669746 non-null  int64  
 4   end_day             669746 non-null  int64  
 5   end_time            669746 non-null  int64  
 6   episode_id          669746 non-null  int64  
 7   event_id            669746 non-null  int64  
 8   state               669746 non-null  object 
 9   state_fips          669746 non-null  int64  
 10  year                669746 non-null  int64  
 11  month_name          669746 non-null  object 
 12  event_type          669746 non-null  object 
 13  cz_type             669746 non-null  object 
 14  cz_fips             669746 non-null  int64  
 15  cz_name             669746 non-nul

We have seneval columns with missing values. Let's check for the columns with missing values.

In [106]:
# Check for missing values in the data
missing_values = df_details.isnull().sum()
print("Missing values in the data:")
missing_values[missing_values > 0].sort_values(ascending=False)

Missing values in the data:


tor_other_cz_name     667758
tor_other_wfo         667758
tor_other_cz_state    667758
tor_other_cz_fips     667758
tor_f_scale           654554
tor_length            654554
tor_width             654554
flood_cause           597354
magnitude_type        416845
magnitude             320801
end_lat               259317
end_lon               259317
begin_range           259317
begin_azimuth         259317
begin_location        259317
end_range             259317
end_azimuth           259317
end_location          259317
begin_lat             259317
begin_lon             259317
event_narrative       140063
damage_property       136418
damage_crops          134127
dtype: int64

Start with `tor_f_scale`, `tor_length`, and `tor_width` columns.
Based on the Data Dictionary provided at the beginning I assume columns like `tor_f_scale`, `tor_length`, and `tor_width` are relevant only for tornado events. It’s expected that these fields would be missing for non-tornado events. Let's prove it.

Now, we can move on to the other columns related to geographical locations (`end_lat`, `end_lon`, `begin_range`, `begin_azimuth`, `begin_location`, `end_range`, `end_azimuth`, `end_location`, `begin_lat`, `begin_lon`). I noticed that they all have the same number of rows with missing values. So, I suspect either all have values or are all null in a row, with no partial cases where some columns are populated while others are not. Let's verify this assumption.

In [79]:
# Select the relevant columns
geo_columns = ['end_lat', 'end_lon', 'begin_range', 'begin_azimuth', 'begin_location', 
               'end_range', 'end_azimuth', 'end_location', 'begin_lat', 'begin_lon']

# Create a subset of just those columns
geo_data = df_details[geo_columns]

# Check if all columns are either fully null or fully populated
all_null_or_all_filled = geo_data.isnull().all(axis=1) | geo_data.notnull().all(axis=1)

# Count the number of rows where some columns are null but others are not
partially_null_rows_count = (~all_null_or_all_filled).sum()

print(f"Number of rows where some columns are null and others are filled: {partially_null_rows_count}")

Number of rows where some columns are null and others are filled: 0


The result indicates that there are no missing values in the tornado-specific columns (`tor_f_scale`, `tor_length`, and `tor_width`) for events classified as tornadoes in the dataset. This confirms my assumption that these fields are only populated for tornado events and should not have missing values within that context.

Thus, we don’t need to worry about these missing values in the broader dataset, as they are valid and expected for non-tornado events. 

Now, let's move on to `tor_other_cz_name`, `tor_other_wfo`, `tor_other_cz_state`, and `tor_other_cz_fips` columns. Based on the Data Dictionary mentioned above, columns like `tor_other_cz_name`, `tor_other_wfo`, `tor_other_cz_state`, and `tor_other_cz_fips` apply only to tornadoes that cross into other geographical areas. If a tornado travels beyond its initial location, these additional columns should contain data.

In [64]:
# Define the columns related to tornadoes that cross geographical areas
tornado_other_location_cols = ['tor_other_cz_name', 'tor_other_wfo', 'tor_other_cz_state', 'tor_other_cz_fips']

# 1. Prove there are only two outcomes: either all four are null or all four are not null
# Create a mask where either all 4 are null or all 4 are not null
all_null_or_not_null = (
    (df_details[tornado_other_location_cols].isnull().all(axis=1)) | 
    (df_details[tornado_other_location_cols].notnull().all(axis=1))
)

# Check if any rows violate this condition
invalid_rows = df_details[~all_null_or_not_null]

if invalid_rows.empty:
    print("There are no cases where only some of the 'tor_other_*' columns are null/non-null. The assumption holds.")
else:
    print("There are cases where only some of the 'tor_other_*' columns are null/non-null:")
    print(invalid_rows)

# 2. Prove that if the 4 columns are null, event_type is not Tornado, and if not null, event_type is Tornado and the original locations differs from the other locations

# First check if event_type is not Tornado when all 4 columns are null
non_tornado_mismatch = df_details[(df_details[tornado_other_location_cols].notnull().all(axis=1)) & (~df_details['event_type'].str.contains('Tornado', na=False))]

# Check if the initial locations are different from the other locations when the 4 columns are not null and event_type is Tornado
location_mismatch = df_details[(df_details[tornado_other_location_cols].notnull().all(axis=1)) & 
                               (df_details['event_type'].str.contains('Tornado', na=False)) &
                               (df_details['cz_fips'] == df_details['tor_other_cz_fips']) &
                                (df_details['cz_name'] == df_details['tor_other_cz_name']) &
                                (df_details['wfo'] == df_details['tor_other_wfo'])]

# Output the results
if non_tornado_mismatch.empty and location_mismatch.empty:
    print("All conditions hold: tornado events have 'tor_other_*' columns populated and locations differ when these columns are populated.")
else:
    if not non_tornado_mismatch.empty:
        print("There are non-tornado events where 'tor_other_*' columns are populated:")
        print(non_tornado_mismatch[['event_type'] + tornado_other_location_cols])
    if not location_mismatch.empty:
        print("There are tornado events where the initial location equals the other location but 'tor_other_*' columns are populated:")
        print(location_mismatch[['event_id', 'event_type', 'cz_name', 'wfo', 'cz_type', 'cz_fips',] + tornado_other_location_cols])

There are no cases where only some of the 'tor_other_*' columns are null/non-null. The assumption holds.
All conditions hold: tornado events have 'tor_other_*' columns populated and locations differ when these columns are populated.


This means that these columns only contain data when a tornado crosses into a new geographical area. If a tornado stays within one boundary, these columns remain null. This is consistent with the data dictionary and the expected behavior of the dataset.

Next, let's check the `flood_cause` column. The `flood_cause` column is expected to be populated only for events classified as floods. Let's verify this assumption.

In [68]:
# Flood-specific check
flood_events = df_details[df_details['event_type'].str.contains('Flood', na=False)]
non_flood_events = df_details[~df_details['event_type'].str.contains('Flood', na=False)]
print(f"Non-flood events with flood_cause filled: {non_flood_events['flood_cause'].notnull().sum()}")

Non-flood events with flood_cause filled: 1409


Most of the flood events will have the `flood_cause` populated. Let's check what `event_type` values are associated with the 1409 exceptions above.

In [70]:
# Display the event types for non-flood events with flood_cause filled
non_flood_events_with_flood_cause = non_flood_events[non_flood_events['flood_cause'].notnull()]

non_flood_events_with_flood_cause['event_type'].value_counts()

event_type
Debris Flow    1409
Name: count, dtype: int64

The result shows that the `flood_cause` column is filled for events that are not classified as floods. In this case, 1409 rows have the `flood_cause` column populated while their associated `event_type` is "Debris Flow".

Debris flow can be influenced by several factors, but it typically occurs as a result of **heavy rainfall**. This could explain why the `flood_cause` column is populated for these events. So, we do not need to handle these missing values in the `flood_cause` column as they are valid entries.

Now, let's check the `magnitude` column. The `magnitude` column is expected to be populated for events classified as wind or hail. Let's verify this assumption.

In [110]:
# Check for missing values in the magnitude column
missing_magnitude = df_details['magnitude'].isnull().sum()
print(f"Number of missing values in 'magnitude': {missing_magnitude}\n")

# Identify event types with missing magnitude
missing_magnitude_events = df_details[df_details['magnitude'].isnull()]['event_type'].value_counts()
print("Event types with missing magnitude:")
missing_magnitude_events

Number of missing values in 'magnitude': 320801

Event types with missing magnitude:


event_type
Flash Flood                   41325
Winter Weather                40129
Winter Storm                  31873
Drought                       31702
Flood                         29658
Heavy Snow                    24229
Tornado                       15192
Heavy Rain                    14871
Heat                          13209
Excessive Heat                11608
Extreme Cold/Wind Chill        9508
Dense Fog                      7529
Cold/Wind Chill                6295
Frost/Freeze                   6087
Blizzard                       5888
High Surf                      4326
Lightning                      3519
Wildfire                       3447
Funnel Cloud                   2959
Tropical Storm                 2664
Ice Storm                      2566
Coastal Flood                  2385
Waterspout                     2088
Debris Flow                    1409
Lake-Effect Snow               1028
Dust Storm                      982
Rip Current                     868
Marine Tropical S

In [111]:
# Check non-null magnitude values and their associated event types
non_null_magnitude_events = df_details[df_details['magnitude'].notnull()]['event_type'].value_counts()
print("Event types with non-null magnitude:")
non_null_magnitude_events

Event types with non-null magnitude:


event_type
Thunderstorm Wind           178272
Hail                         95763
High Wind                    38392
Marine Thunderstorm Wind     24362
Strong Wind                  11197
Marine High Wind               602
Marine Hail                    281
Marine Strong Wind              76
Name: count, dtype: int64

The majority of the event types associated with missing `magnitude` values are not typically associated with specific magnitudes, such as "Flash Flood", "Winter Weather", "Winter Storm" and other while the event types that have non-null values for `magnitude` primarily include "Thunderstorm Wind", "Hail", and "High Wind". This aligns with the definition provided in the Data Dictionary. Thus, the missing values in the `magnitude` column are expected and do not require any further action.

Now, let's check `magnitude_type` column. The `magnitude_type` column is expected to be populated for events classified as wind and has an associated value in the `magnitude` column. Let's verify this assumption.

In [77]:
# Filter rows where magnitude_type is not null
magnitude_type_non_null = df_details[df_details['magnitude_type'].notnull()]

# Check if magnitude_type rows have non-null values in magnitude (Condition 1)
magnitude_missing = magnitude_type_non_null[magnitude_type_non_null['magnitude'].isnull()]

# Define wind-related event types (Condition 2)
wind_event_types = ['Thunderstorm Wind', 'High Wind', 'Strong Wind', 'Marine Thunderstorm Wind', 
                    'Marine High Wind', 'Marine Strong Wind']

# Check if magnitude_type rows have wind-related event types (Condition 2)
non_wind_events = magnitude_type_non_null[~magnitude_type_non_null['event_type'].isin(wind_event_types)]

# Output results
print(f"Rows where magnitude_type is non-null but magnitude is null: {len(magnitude_missing)}")
print(f"Event types with non-null magnitude_type but not wind-related: {len(non_wind_events)}")

# If any violations are found, display the problematic rows
if not magnitude_missing.empty:
    print("Rows with magnitude_type but missing magnitude:")
    print(magnitude_missing[['event_type', 'magnitude_type', 'magnitude']])

if not non_wind_events.empty:
    print("Rows with magnitude_type but non-wind event type:")
    print(non_wind_events[['event_type', 'magnitude_type']])

Rows where magnitude_type is non-null but magnitude is null: 0
Event types with non-null magnitude_type but not wind-related: 0


The `magnitude_type` field is correctly populated only for wind-related events and always has an associated value in the `magnitude` column. No cases were found where `magnitude_type` was non-null without a corresponding value in `magnitude`, or where `magnitude_type` was filled for non-wind events. This confirms that the field is used appropriately in the dataset. So no further action is required.

Now, let's check the columns related to geographical locations, such as `end_lat`, `end_lon`, `begin_range`, `begin_azimuth`, `begin_location`, `end_range`, `end_azimuth`, `end_location`, `begin_lat`, and `begin_lon`. I noticed they have the same number of rows with missing values. Let's verify if they are all null or all populated in each row.

In [80]:
# Select the relevant columns
geo_columns = ['end_lat', 'end_lon', 'begin_range', 'begin_azimuth', 'begin_location', 
               'end_range', 'end_azimuth', 'end_location', 'begin_lat', 'begin_lon']

# Create a subset of just those columns
geo_data = df_details[geo_columns]

# Check if all columns are either fully null or fully populated
all_null_or_all_filled = geo_data.isnull().all(axis=1) | geo_data.notnull().all(axis=1)

# Count the number of rows where some columns are null but others are not
partially_null_rows_count = (~all_null_or_all_filled).sum()

print(f"Number of rows where some columns are null and others are filled: {partially_null_rows_count}")

Number of rows where some columns are null and others are filled: 0


So, that confirms my theory. These columns are either all being populated or all null in a row, with no partial cases where some columns are populated while others are not. This leads to me to believe that these columns are not always required for every event in the dataset. It makes sense that events like droughts or heat waves, for example, may not have clearly defined start or end geographical locations because they can affect broad regions over time rather than specific points.

We can check if the missing geographical data is associated with specific types of events that typically don’t have a clear starting or ending location. 

In [112]:
# Filter rows where the geographical data is missing
missing_geo_data = df_details[geo_data.isnull().all(axis=1)]

# Count the event types for these rows
missing_geo_event_types = missing_geo_data['event_type'].value_counts()

print("Event types associated with missing geographical data:")
missing_geo_event_types

Event types associated with missing geographical data:


event_type
Winter Weather                40129
High Wind                     38392
Winter Storm                  31873
Drought                       31702
Heavy Snow                    24229
Heat                          13209
Excessive Heat                11608
Strong Wind                   11197
Extreme Cold/Wind Chill        9508
Dense Fog                      7529
Cold/Wind Chill                6295
Frost/Freeze                   6087
Blizzard                       5888
High Surf                      4326
Wildfire                       3447
Tropical Storm                 2664
Ice Storm                      2566
Coastal Flood                  2385
Lake-Effect Snow               1028
Dust Storm                      982
Rip Current                     868
Marine Tropical Storm           506
Storm Surge/Tide                443
Astronomical Low Tide           353
Avalanche                       348
Sleet                           321
Lakeshore Flood                 315
Hurricane        

This result supports the idea that the missing geographical data (latitude, longitude, range, azimuth, and location) is associated with events that typically don’t have well-defined start and end points.

Given this context, we can conclude that the missing values in these columns are valid and do not require further action.

Now, let's check `event_narrative` column. The `event_narrative` column provides descriptive details of the individual event. Let's verify if the missing values in this column are associated with specific event types.

In [113]:
# Check for missing values in the event_narrative column
missing_narrative_events = df_details[df_details['event_narrative'].isnull()]

# Count the event types in these rows
event_narrative_counts = missing_narrative_events['event_type'].value_counts()

print("Event types with missing event_narrative:")
event_narrative_counts

Event types with missing event_narrative:


event_type
Hail                        50120
Winter Weather              10362
Drought                     10090
Thunderstorm Wind            8795
Winter Storm                 7033
High Wind                    6845
Heat                         5986
Excessive Heat               4923
Heavy Snow                   4547
Strong Wind                  4363
High Surf                    3844
Extreme Cold/Wind Chill      3549
Dense Fog                    3347
Cold/Wind Chill              3024
Marine Thunderstorm Wind     2499
Frost/Freeze                 2373
Blizzard                     2097
Heavy Rain                   1374
Flood                         886
Wildfire                      787
Lake-Effect Snow              566
Ice Storm                     487
Funnel Cloud                  470
Tropical Storm                363
Flash Flood                   244
Waterspout                    113
Lakeshore Flood               109
Coastal Flood                 105
Marine Tropical Storm          94
Ast

In our analysis, we observed a significant number of missing values in the `event_narrative` column. Given that the narrative primarily provides additional context rather than contributing to the quantitative analysis, we will not be filling in these missing values. Instead, we will retain the existing data as is, ensuring that our analysis remains focused on quantifiable metrics. Furthermore, since the `episode_narrative` column has no missing values, it may serve as a useful supplementary reference, although it will not be included in our main analysis.



For the `damage_property` column, we will impute the missing values, which account for approximately 20.38% of the dataset. Given this high percentage, removing these missing rows could significantly bias our analysis and lead to a loss of valuable information. We will use averages calculated for combinations of `state`, `event_type`, and `year` for most cases, but where all values in a group are missing, we will resort to the overall average. This mixed approach allows us to consider regional and event-specific variations while still ensuring that we have a method for imputing values when necessary. By leveraging both types of averages, we aim to provide a more accurate estimate than relying solely on an overall average.

In [143]:
# Convert 'damage_property' to numeric
def convert_damage(value):
    if isinstance(value, str):
        if 'K' in value:
            return float(value.replace('K', '').strip()) * 1_000
        elif 'M' in value:
            return float(value.replace('M', '').strip()) * 1_000_000
        elif 'B' in value:
            return float(value.replace('B', '').strip()) * 1_000_000_000
    return np.nan

# Apply the conversion
df_details['damage_property_numeric'] = df_details['damage_property'].apply(convert_damage)

# Calculate average damage_property_numeric by state, event_type, and year
averages = df_details.groupby(['state', 'event_type', 'year'])['damage_property_numeric'].mean().reset_index()
averages.rename(columns={'damage_property_numeric': 'average_damage_property'}, inplace=True)

# Merge the averages back to the original DataFrame
df_details = df_details.merge(averages, on=['state', 'event_type', 'year'], how='left')

# Impute missing values with the calculated averages
df_details['imputed_damage_property'] = df_details['damage_property_numeric'].fillna(df_details['average_damage_property'])

# Fill remaining missing values with the overall average
overall_avg_damage = df_details['damage_property_numeric'].mean()
df_details['imputed_damage_property'] = df_details['imputed_damage_property'].fillna(overall_avg_damage)

# Check for remaining missing values
remaining_missing = df_details['imputed_damage_property'].isnull().sum()
print(f"Remaining missing values: {remaining_missing}")

Remaining missing values: 0


In [150]:
df_details[df_details['damage_property'].isnull()][['year', 'state', 'event_type', 'damage_property', 'damage_property_numeric', 'imputed_damage_property']]

year           state                event_type damage_property  \
24      2021            IOWA                   Tornado             NaN   
92      2021            IOWA                   Tornado             NaN   
139     2021       LOUISIANA         Thunderstorm Wind             NaN   
146     2021     MISSISSIPPI         Thunderstorm Wind             NaN   
147     2021     MISSISSIPPI         Thunderstorm Wind             NaN   
...      ...             ...                       ...             ...   
669738  2017         FLORIDA          Storm Surge/Tide             NaN   
669739  2017         FLORIDA          Storm Surge/Tide             NaN   
669741  2017  GULF OF MEXICO  Marine Hurricane/Typhoon             NaN   
669742  2017  GULF OF MEXICO  Marine Hurricane/Typhoon             NaN   
669743  2017         FLORIDA          Storm Surge/Tide             NaN   

        damage_property_numeric  imputed_damage_property  
24                          NaN             1.359400e+05  
92                          NaN             1.359400e+05  
139                         NaN             4.477059e+03  
146                         NaN             2.150181e+04  
147                         NaN             2.150181e+04  
...                         ...                      ...  
669738                      NaN             1.557692e+06  
669739                      NaN             1.557692e+06  
669741                      NaN             0.000000e+00  
669742                      NaN             0.000000e+00  
669743                      NaN             1.557692e+06  

[136418 rows x 6 columns]

## Appendix

In [138]:
# Calculate percentage of missing values by state
missing_by_state = df_details['damage_property'].isnull().groupby(df_details['state']).mean() * 100

missing_by_state.sort_values(ascending=False)

state
MAINE                   62.568606
ATLANTIC NORTH          60.294638
GUAM                    54.128440
SOUTH CAROLINA          52.813196
DISTRICT OF COLUMBIA    45.622120
                          ...    
LAKE ST CLAIR            0.719424
ARIZONA                  0.402510
GULF OF ALASKA           0.000000
HAWAII WATERS            0.000000
GUAM WATERS              0.000000
Name: damage_property, Length: 69, dtype: float64

In [136]:
# Calculate percentage of missing values by event_type
missing_by_event_type = df_details['damage_property'].isnull().groupby(df_details['event_type']).mean() * 100

missing_by_event_type.sort_values(ascending=False)

event_type
Ice Storm                     36.204209
Marine Dense Fog              33.333333
Tropical Depression           33.000000
Hurricane (Typhoon)           31.428571
Blizzard                      31.266984
Drought                       30.505962
Dense Fog                     30.203214
Coastal Flood                 29.769392
Hail                          29.275399
Extreme Cold/Wind Chill       29.196466
Marine Thunderstorm Wind      27.858961
Cold/Wind Chill               27.418586
Winter Weather                26.260310
Tropical Storm                24.174174
Storm Surge/Tide              22.121896
Winter Storm                  22.062561
Thunderstorm Wind             21.104829
Frost/Freeze                  20.174142
Heavy Snow                    20.124644
High Wind                     19.590019
Heat                          19.191460
Marine Hail                   18.149466
Sleet                         18.068536
Wildfire                      17.986655
Dust Storm                   

In [137]:
# Calculate percentage of missing values by year
missing_by_year = df_details['damage_property'].isnull().groupby(df_details['year']).mean() * 100

missing_by_year.sort_values(ascending=False)

year
2023    23.618144
2022    23.453911
2024    21.647346
2020    20.989246
2016    20.878493
2021    20.438515
2019    19.585624
2018    19.524060
2017    18.569500
2015    17.393407
2014    16.965111
Name: damage_property, dtype: float64

In [145]:
# Group by year, state, and event_type, then calculate the percentage of missing damage_property
missing_groups = df_details.groupby(['year', 'state', 'event_type'])['damage_property'].apply(lambda x: x.isna().mean()).reset_index()

# Filter for groups where the missing percentage is 100%
missing_100_percent = missing_groups[missing_groups['damage_property'] == 1.0]

# Display the result
missing_100_percent

year          state             event_type  damage_property
10     2014        ALABAMA                  Sleet              1.0
26     2014         ALASKA       Storm Surge/Tide              1.0
54     2014       ARKANSAS           Frost/Freeze              1.0
123    2014    CONNECTICUT        Cold/Wind Chill              1.0
163    2014        FLORIDA  Astronomical Low Tide              1.0
...     ...            ...                    ...              ...
10689  2024        VERMONT             Heavy Snow              1.0
10721  2024     WASHINGTON                   Heat              1.0
10722  2024     WASHINGTON             Heavy Rain              1.0
10745  2024  WEST VIRGINIA               Wildfire              1.0
10764  2024        WYOMING               Blizzard              1.0

[761 rows x 4 columns]

### Outline for Python Notebook: Storm Event Data Analysis

#### 1. **Ask**
   - **Define the Problem**
     - Identify the key questions: 
       - What types of storms are most common in the US?
       - Where do most storms originate?
       - How do storms impact communities (injuries, deaths, damage)?
   - **Confirm Stakeholder Expectations**
     - Define what stakeholders expect from the analysis (e.g., actionable insights, visualizations).

#### 2. **Prepare**
   - **Collect Data**
     - Load the storm event dataset using `pandas`.
     - Explore the dataset structure using `info()`, `head()`, and `describe()`.
   - **Store Data**
     - Save any preliminary data transformations or filtered datasets for later use.

#### 3. **Process**
   - **Clean the Data**
     - Handle missing values and outliers.
     - Convert data types as necessary (e.g., date parsing).
   - **Transform Data**
     - Standardize column names to snake_case.
     - Filter data for specific storm events or time periods if needed.
     - Create new columns for analysis (e.g., total damage).

#### 4. **Analyze**
   - **Descriptive Analysis**
     - Count unique storm events by type.
     - Calculate total injuries, deaths, and property damage.
   - **Geospatial Analysis**
     - Use latitude and longitude data to plot storm origins on a map.
     - Identify regions most affected by specific storm types.
   - **Trends and Patterns**
     - Analyze trends over time (e.g., increase in storm frequency).

#### 5. **Share**
   - **Visualizations**
     - Create graphs and maps to illustrate key findings.
     - Use interactive maps for better engagement.
   - **Interpret Results**
     - Summarize insights from the data (e.g., most common storms, geographic hotspots).

#### 6. **Act**
   - **Recommendations**
     - Provide actionable insights based on findings.
     - Suggest further areas for research or monitoring based on trends.

### Notes
- Ensure to document each step with comments and markdown cells to explain your thought process and findings.
- Include visualizations at appropriate points to enhance understanding.
